// react-tag-input-component/dist/react-tag-input-component.esm.js
import { css, setup } from 'goober';
import React, { useState, useEffect } from 'react';

/**
 * A minimal utility to combine classes
 *
 * @export
 * @param {(string[] | string)} obj
 * @returns {string}
 */
function cc() {
  for (var _len = arguments.length, obj = new Array(_len), _key = 0; _key < _len; _key++) {
    obj[_key] = arguments[_key];
  }

  return obj.join(" ");
}

var tagStyles = /*#__PURE__*/css({
  alignItems: "center",
  background: "var(--rti-tag)",
  borderRadius: "var(--rti-radius)",
  display: "inline-flex",
  justifyContent: "center",
  paddingLeft: "var(--rti-s)",
  button: {
    background: "none",
    border: 0,
    borderRadius: "50%",
    cursor: "pointer",
    lineHeight: "inherit",
    padding: "0 var(--rti-s)",
    "&:hover": {
      color: "var(--rti-tag-remove)"
    }
  }
});
function Tag(_ref) {
  var text = _ref.text,
      remove = _ref.remove;

  var handleOnRemove = function handleOnRemove(e) {
    e.stopPropagation();
    remove(text);
  };

  return React.createElement("span", {
    className: cc("rti--tag", tagStyles)
  }, React.createElement("span", null, text), React.createElement("button", {
    type: "button",
    onClick: handleOnRemove,
    "aria-label": "remove " + text
  }, "\u2715"));
}

setup(React.createElement);
var RTIContainer = /*#__PURE__*/css({
  "--rtiBg": "#fff",
  "--rtiBorder": "#ccc",
  "--rtiMain": "#3182ce",
  "--rtiRadius": "0.375rem",
  "--rtiS": "0.5rem",
  "--rtiTag": "#edf2f7",
  "--rtiTagRemove": "#e53e3e",
  "*": {
    boxSizing: "border-box",
    transition: "all 0.2s ease"
  },
  alignItems: "center",
  bg: "var(--rti-bg)",
  border: "1px solid var(--rti-border)",
  borderRadius: "var(--rti-radius)",
  display: "flex",
  flexWrap: "wrap",
  gap: "var(--rti-s)",
  lineHeight: 1.4,
  padding: "var(--rti-s)",
  "&:focus-within": {
    borderColor: "var(--rti-main)",
    boxShadow: "var(--rti-main) 0px 0px 0px 1px"
  }
});
var RTIInput = /*#__PURE__*/css({
  border: 0,
  outline: 0,
  fontSize: "inherit",
  lineHeight: "inherit",
  width: "50%"
});
var defaultSeprators = ["Enter"];
var TagsInput = function TagsInput(_ref) {
  var name = _ref.name,
      placeHolder = _ref.placeHolder,
      value = _ref.value,
      onChange = _ref.onChange,
      onBlur = _ref.onBlur,
      seprators = _ref.seprators,
      onExisting = _ref.onExisting,
      onRemoved = _ref.onRemoved;

  var _useState = useState(value || []),
      tags = _useState[0],
      setTags = _useState[1];

  useEffect(function () {
    onChange && onChange(tags);
  }, [tags]);

  var handleOnKeyUp = function handleOnKeyUp(e) {
    e.stopPropagation();
    var text = e.target.value;

    if (e.key === "Backspace" && tags.length && !text) {
      setTags(tags.slice(0, -1));
    }

    if (text && (seprators || defaultSeprators).includes(e.key)) {
      if (tags.includes(text)) {
        onExisting && onExisting(text);
        return;
      }

      setTags([].concat(tags, [text]));
      e.target.value = "";
      e.preventDefault();
    }
  };

  var onTagRemove = function onTagRemove(text) {
    setTags(tags.filter(function (tag) {
      return tag !== text;
    }));
    onRemoved && onRemoved(text);
  };

  return React.createElement("div", {
    "aria-labelledby": name,
    className: cc("rti--container", RTIContainer)
  }, tags.map(function (tag) {
    return React.createElement(Tag, {
      key: tag,
      text: tag,
      remove: onTagRemove
    });
  }), React.createElement("input", {
    className: cc("rti--input", RTIInput),
    type: "text",
    name: name,
    placeholder: placeHolder,
    onKeyDown: handleOnKeyUp,
    onBlur: onBlur
  }));
};

export { TagsInput };
